File: src/csoclassify/record.py
#------------------------------------------------------------------
# 분류 레코드 읽기 도우미 — 여러 세대의 모양을 한 자리에서 흡수한다
#=> 레코드 모양은 세 번 바뀌었다.
#
#   (1세대) 최상위 grade + labels:{security:{value,…}, doctype:{…}} + signals
#   (2세대, 2026-09-10 오전) security:{grade,…,signals} + doctype:{values,…} + meta
#   (3세대, 2026-09-10 오후) 판정은 앞에, 근거는 why 안에
#
#       {file, hash, doc_id, doc_id_source,
#        grade: "C",                       ← 등급 판정(한 벌뿐)
#        doctype: ["DC_002_002", …],       ← 문서분류 판정(dc_id 목록)
#        why: {security: {confidence, method, decided_by, seed_eligible, signals},
#              doctype:  {values, [strategy], [truncated], [conflicts], [embed]}},
#        meta: {ts}}
#
#   3세대의 뜻은 "무엇을 → 어떻게 됐나 → 왜 → 장부" 다. 판정을 맨 앞에 한 벌만
#   두고, 그 판정을 뒷받침하는 것은 전부 why 안에 둔다. --simple 출력이 앞의
#   두 묶음만 떼어낸 것과 같은 모양이라, 축약본과 전체가 같은 낱말을 쓴다.
#
#   [왜 별도 모듈인가] 레코드를 읽는 곳이 엔진·CLI·검토 화면에 흩어져 있다.
#   각자 "새 모양이면 …, 옛 모양이면 …" 을 적으면 그 조건문이 스무 군데로
#   퍼지고, 한 군데만 빠뜨려도 옛 결과 파일이 조용히 안 보이게 된다.
#   그래서 흡수를 이 파일 하나에 모은다.
#
#   [옛 파일을 왜 계속 읽나] 이 도구는 "언제 무엇을 왜 그렇게 판정했는가"를
#   쌓아 두는 것이 일이다. 형식을 바꿨다고 그동안 쌓인 판정을 못 읽게 되면
#   도구의 존재 이유가 사라진다.
#------------------------------------------------------------------


#------------------------------------------------------------------
# 옛 labels 껍데기 안전하게 꺼내기
#=> seed 파일(class_seed.jsonl)도 labels 라는 이름의 칸을 쓰는데 모양이 다르다
#   (거기서는 labels.security 가 등급 '문자열'이다). 결과 레코드로 착각하고
#   .get() 을 부르면 AttributeError 로 터지므로, dict 일 때만 통과시킨다.
#
# -in: rec = 아무 dict
#
# -out: dict = rec["labels"] 가 dict 이면 그것, 아니면 빈 dict
# -out: error = 없음
#------------------------------------------------------------------
def _labels(rec):
    lab = rec.get("labels")
    return lab if isinstance(lab, dict) else {}


#------------------------------------------------------------------
# 근거 묶음(why) 꺼내기
#=> 3세대에만 있는 칸이다. 없으면 빈 dict 를 돌려, 부르는 쪽이 옛 세대 자리를
#   이어서 보게 한다.
#
# -in: rec = 분류 레코드
#
# -out: dict = rec["why"] 가 dict 이면 그것, 아니면 빈 dict
# -out: error = 없음
#------------------------------------------------------------------
def _why(rec):
    w = rec.get("why")
    return w if isinstance(w, dict) else {}


#------------------------------------------------------------------
# 최종 등급
#=> 3세대·1세대는 최상위 grade 가 곧 답이다(같은 이름을 그대로 되살렸다).
#   2세대만 security.grade 안에 있었다.
#
# -in: rec = 분류 레코드
#
# -out: "C"|"S"|"O"|None — 아무 신호도 못 정했으면 None(보류)
# -out: error = 없음
#------------------------------------------------------------------
def grade_of(rec):
    rec = rec or {}
    if "grade" in rec:
        return rec.get("grade")
    sec = rec.get("security")
    if isinstance(sec, dict):
        return sec.get("grade")
    old = _labels(rec).get("security")
    return old.get("value") if isinstance(old, dict) else None


#------------------------------------------------------------------
# 등급 축의 판정·근거 묶음
#=> 3세대는 why.security 다. 편의를 위해 grade 를 함께 채워 돌려준다 —
#   읽는 쪽 대부분이 "등급과 확신을 같이" 보고 싶어 하기 때문이다(레코드에
#   실제로 적히는 모양과는 다르다. 레코드에는 grade 가 맨 앞에 한 벌뿐이다).
#
# -in: rec = 분류 레코드(어느 세대든)
#
# -out: dict = {grade, confidence, method, decided_by, seed_eligible, signals}
# -out: error = 없음
#------------------------------------------------------------------
def security_of(rec):
    rec = rec or {}
    sec = _why(rec).get("security")
    if not isinstance(sec, dict):
        sec = rec.get("security")
    if isinstance(sec, dict):
        out = dict(sec)
        out["grade"] = grade_of(rec)
        return out
    # ── 1세대 ────────────────────────────────────────────────
    old = _labels(rec).get("security")
    old = old if isinstance(old, dict) else {}
    return {
        "grade": grade_of(rec),
        "confidence": old.get("confidence", rec.get("confidence", 0.0)),
        "method": old.get("method", rec.get("method", "unclassified")),
        "decided_by": old.get("decided_by", rec.get("decided_by") or []),
        "seed_eligible": rec.get("seed_eligible", False),
        "signals": rec.get("signals") or {},
    }


#------------------------------------------------------------------
# 문서분류체계 축의 판정·근거 묶음
#=> [축을 안 쓰는 배포와 구분] 이 축을 아예 안 돌린 배포는 칸 자체가 없고,
#   돌렸는데 후보를 못 찾은 문서는 values 가 빈 목록이다. 이 둘은 뜻이
#   완전히 다르므로(전자는 "안 봄", 후자는 "봤는데 없음") None 과 dict 로
#   갈라서 돌려준다 — 합치면 "왜 비었는지"를 영영 알 수 없다.
#
#   3세대의 최상위 doctype 은 dc_id **배열**이라 여기서 찾는 대상이 아니다.
#   상세는 why.doctype 에 있고, 배열은 그것에서 뽑은 값이다.
#
# -in: rec = 분류 레코드
#
# -out: dict|None = {values, [strategy], [truncated], [conflicts], [embed]}
#                   또는 축을 안 돌렸으면 None
# -out: error = 없음
#------------------------------------------------------------------
def doctype_of(rec):
    rec = rec or {}
    dt = _why(rec).get("doctype")
    if isinstance(dt, dict):
        return dt
    dt = rec.get("doctype")          # 2세대는 최상위가 객체였다
    if isinstance(dt, dict):
        return dt
    old = _labels(rec).get("doctype")
    return old if isinstance(old, dict) else None


#------------------------------------------------------------------
# 문서분류체계 후보 목록
#=> 축을 안 돌렸든 후보가 없든 '빈 목록'으로 통일한다. 그 둘을 갈라야 하는
#   곳은 doctype_of() 로 None 여부를 직접 보면 된다.
#
# -in: rec = 분류 레코드
#
# -out: list = 후보 dict 목록(없으면 빈 목록)
# -out: error = 없음
#------------------------------------------------------------------
def doctype_values(rec):
    dt = doctype_of(rec)
    return (dt.get("values") or []) if dt else []


#------------------------------------------------------------------
# 확정된 문서분류 dc_id 목록
#=> 3세대 레코드는 이 목록이 최상위 doctype 에 이미 적혀 있다(판정을 앞에
#   두자는 뜻). 그 밖의 세대는 후보에서 뽑아 만든다.
#
# -in: rec = 분류 레코드
#
# -out: list = dc_id 문자열 목록
# -out: error = 없음
#------------------------------------------------------------------
def doctype_ids(rec):
    rec = rec or {}
    dt = rec.get("doctype")
    if isinstance(dt, list):
        return list(dt)
    return [v.get("dc_id") for v in doctype_values(rec) if v.get("dc_id")]


#------------------------------------------------------------------
# 부가 정보(시각·버전) 꺼내기
#=> 3세대부터 규칙셋 버전 세 칸은 레코드마다 적지 않고 결과 파일 맨 앞의
#   실행 헤더({"run": …})에 한 번만 적는다. 레코드에 남는 것은 ts 뿐이다.
#   옛 세대 파일에는 여전히 레코드마다 들어 있어, 있으면 그대로 읽는다.
#
# -in: rec = 분류 레코드
#
# -out: dict = {ts[, rule_version, taxonomy_version, doctype_rule_version]}
# -out: error = 없음
#------------------------------------------------------------------
def meta_of(rec):
    rec = rec or {}
    meta = rec.get("meta")
    if isinstance(meta, dict):
        return meta
    return {k: rec[k] for k in
            ("rule_version", "taxonomy_version", "doctype_rule_version", "ts")
            if k in rec}

File: Rust/src/record.rs
//! 분류 레코드 읽기 도우미 — 새 모양과 옛 모양을 한 자리에서 흡수한다.
//!
//! 2026-09-10 부터 레코드가 축별로 평평해졌다.
//!
//! ```text
//! 전:  {grade, confidence, method, decided_by, seed_eligible,
//!       doctype:[dc_id…], labels:{security:{value,…}, doctype:{…}},
//!       signals:{…}, rule_version, ts}
//! 후:  {security:{grade, confidence, method, decided_by, seed_eligible, signals:{…}},
//!       doctype:{status, strategy, values, …},
//!       meta:{rule_version, taxonomy_version, doctype_rule_version, ts}}
//! ```
//!
//! 같은 값을 'labels 안'과 '최상위'에 두 벌 적던 것을 한 벌로 줄이고,
//! 등급(security)과 문서분류체계(doctype)를 나란한 두 칸으로 갈랐다.
//!
//! [왜 별도 모듈인가] 레코드를 읽는 곳이 여기저기 흩어져 있다. 각자
//! "새 모양이면 …, 옛 모양이면 …" 을 적으면 그 조건문이 스무 군데로 퍼지고,
//! 한 군데만 빠뜨려도 옛 결과 파일이 조용히 안 보이게 된다.
//! 파이썬 판의 `csoclassify/record.py` 와 짝이다 — 한쪽만 고치면 두 판이 갈린다.

use serde_json::Value;

/// 옛 `labels` 껍데기를 안전하게 꺼낸다.
/// seed 파일(class_seed.jsonl)도 `labels` 라는 이름의 칸을 쓰는데 모양이 다르다
/// (거기서는 `labels.security` 가 등급 '문자열'이다). 객체일 때만 통과시킨다.
fn labels(rec: &Value) -> Option<&Value> {
    rec.get("labels").filter(|l| l.is_object())
}

/// 근거 묶음(`why`). 3세대에만 있는 칸이다.
fn why(rec: &Value) -> Option<&Value> {
    rec.get("why").filter(|w| w.is_object())
}

/// 등급 축의 근거 묶음. 3세대는 `why.security`, 2세대는 최상위 `security`.
pub fn security<'a>(rec: &'a Value) -> Option<&'a Value> {
    why(rec).and_then(|w| w.get("security")).filter(|s| s.is_object())
        .or_else(|| rec.get("security").filter(|s| s.is_object()))
}

/// 최종 등급("C"/"S"/"O"). 아무 신호도 못 정했으면 `None`(보류).
/// 새 모양 → 옛 `labels.security.value` → 옛 최상위 `grade` 순으로 본다.
pub fn grade_of(rec: &Value) -> Option<&str> {
    // 3세대·1세대는 최상위 grade 가 곧 답이다(같은 이름을 그대로 되살렸다).
    if rec.get("grade").is_some() {
        return rec.get("grade").and_then(|g| g.as_str());
    }
    // 2세대만 security.grade 안에 있었다.
    if let Some(s) = rec.get("security").filter(|s| s.is_object()) {
        return s.get("grade").and_then(|g| g.as_str());
    }
    labels(rec).and_then(|l| l.pointer("/security/value")).and_then(|g| g.as_str())
}

/// 최종 등급을 만든 신호 이름들(`decided_by`).
pub fn decided_by<'a>(rec: &'a Value) -> Option<&'a Value> {
    security(rec).and_then(|s| s.get("decided_by"))
        .or_else(|| labels(rec).and_then(|l| l.pointer("/security/decided_by")))
        .or_else(|| rec.get("decided_by"))
}

/// 신호별 근거(`security.signals`). 옛 레코드는 최상위 `signals`.
pub fn signals<'a>(rec: &'a Value) -> Option<&'a Value> {
    security(rec).and_then(|s| s.get("signals")).filter(|s| s.is_object())
        .or_else(|| rec.get("signals").filter(|s| s.is_object()))
}

/// 문서분류체계 축.
///
/// [축을 안 쓰는 배포와 구분] 이 축을 아예 안 돌린 배포는 칸 자체가 없고,
/// 돌렸는데 후보를 못 찾은 문서는 `values` 가 빈 목록이다. 이 둘은 뜻이 완전히
/// 다르므로(전자는 "안 봄", 후자는 "봤는데 없음") `None` 과 `Some` 으로 가른다.
/// 옛 레코드의 최상위 `doctype` 은 dc_id **배열**이라 객체 검사가 둘을 갈라 준다.
pub fn doctype<'a>(rec: &'a Value) -> Option<&'a Value> {
    if let Some(d) = why(rec).and_then(|w| w.get("doctype")).filter(|d| d.is_object()) {
        return Some(d);
    }
    if let Some(d) = rec.get("doctype").filter(|d| d.is_object()) {
        return Some(d);          // 2세대는 최상위가 객체였다
    }
    labels(rec).and_then(|l| l.get("doctype")).filter(|d| d.is_object())
}

/// 확정된 문서분류 dc_id 목록. 예전에는 이 목록을 최상위 `doctype` 칸에 따로
/// 적어 뒀는데, `values` 에서 기계적으로 나오는 값이라 두 번 적는 것이었다.
pub fn doctype_ids(rec: &Value) -> Vec<Value> {
    // 옛 레코드의 최상위 doctype 은 이미 dc_id 목록이다.
    if let Some(a) = rec.get("doctype").and_then(|d| d.as_array()) {
        return a.clone();
    }
    doctype(rec)
        .and_then(|d| d.get("values"))
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|v| v.get("dc_id").cloned()).collect())
        .unwrap_or_default()
}

/// seed 승격 후보 표식(`security.seed_eligible`).
///
/// [왜 접근자가 필요한가] 2026-09-10 평탄화로 이 칸이 최상위에서 `security`
/// 안으로 들어갔다. 그런데 읽는 쪽 두 군데(1차 임베딩 판정·`--propagate` 의 내부
/// seed 수집)가 여전히 최상위를 보고 있어 **언제나 false** 였다. 오류가 안 나고
/// 값만 조용히 뒤집히는 종류라, 규칙이 고신뢰로 찍어 준 문서가 임베딩에서
/// 통째로 빠지고 업무분류 2단계가 사라지는데도 아무도 못 알아챘다.
/// 파이썬 `record.security_of(rec).get("seed_eligible")` 과 같은 순서로 본다.
pub fn seed_eligible_of(rec: &Value) -> bool {
    if let Some(s) = security(rec) {
        return s.get("seed_eligible").and_then(|b| b.as_bool()).unwrap_or(false);
    }
    // 옛 레코드는 최상위에 있었다(파이썬 security_of 의 옛 모양 가지와 같다).
    rec.get("seed_eligible").and_then(|b| b.as_bool()).unwrap_or(false)
}
